src/main/cpp/src/get_json_object.hpp

```diff
@@ -19,20 +19,270 @@
 #include <cudf/strings/string_view.hpp>
 #include <cudf/strings/strings_column_view.hpp>
 
+#include <thrust/optional.h>
+
 #include <memory>
 
 namespace spark_rapids_jni {
 
+namespace detail {
+
+/**
+ * path instruction type
+ */
+enum class path_instruction_type { SUBSCRIPT, WILDCARD, KEY, INDEX, NAMED };
+
 /**
- * Extracts json object from a json string based on json path specified, and
- * returns json string of the extracted json object. It will return null if the
- * input json string is invalid.
+ * path instruction
  */
+struct path_instruction {
+  CUDF_HOST_DEVICE inline path_instruction(path_instruction_type _type) : type(_type) {}
+
+  path_instruction_type type;
+
+  // used when type is named type
+  cudf::string_view name{"", 0};
+
+  // used when type is index
+  long index{-1};
+};
+
+class JsonPathParser {
+ public:
+  std::vector<path_instruction> instructions;
+
+  thrust::optional<path_instruction> current_instruction = thrust::nullopt;
+
+  size_t current_start = 0;
+  size_t current_len   = 0;
+
+  bool is_digit(char c) { return c >= '0' && c <= '9'; }
+
+  thrust::optional<std::vector<path_instruction>> parse(const std::string& str)
+  {
+    instructions.clear();
+    current_start = 0;
+    current_len   = 0;
+    size_t pos    = 0;
+    if (str.size() == 2) {
+      // the while loop below will not run if str.size() == 2
+      // and there is no valid path
+      return thrust::nullopt;
+    }
+    if (str[pos++] != '$') { return thrust::nullopt; }
+    bool in_brackets = false;
+    bool in_quotes   = false;
+    if (str[pos] == '[') { in_brackets = true; }
+    if (str[pos] == '\'') {
+      // .'name' is invalid
+      return thrust::nullopt;
+    }
+    while (pos < str.size() - 1) {
+      char from = str[pos];
+      char to   = str[pos + 1];
+      // std::cout << "from: " << from << " to: " << to << " " << in_brackets << " " << in_quotes <<
```

Comment:
Can we remove the debug logging, not just comment it out. Or put it under an `#ifdef`.